### DISCUSSION TITLE
Adding custom response headers and customize cookie settings

### DISCUSSION THREAD

**Discussion by teunoknmi** [ckan/ckan#7067] [Q&A] [answered]
We have a requirement to add some headers to the CKAN response and change the settings on some of the cookies. 
Is there a part in the docs somewhere on how to do this on a low level (I could not find it)? I am struggling on how to do this. Especially the header part I have not found. 

I tryed something like:
```
class CustomizationPlugin(plugins.SingletonPlugin):
    plugins.implements(plugins.IMiddleware)

   def make_middleware(self, app, config):
        log.error("failing")
        @app.after_request
        def my_after_request(response):
            log.error("hello from after request")
            response.headers['x-testing'] = 'apple'
            return response

        return app
```
but this does not seem to work.
We are running version 2.9.

Thanks in advance.

**Comment by teunoknmi** [comment]
```
class CustomizationPlugin(plugins.SingletonPlugin):
    plugins.implements(plugins.IMiddleware)

   def make_middleware(self, app, config):
        @app.after_request
        def my_after_request(response):
             response.headers["X-Content-Type-Options"] = "nosniff"
            return response

        return app

   def make_error_log_middleware(self, app, config):
        return app
```

I have the headers working now (probably a docker caching issue from my side.

I am still struggeling on how to change the properties of the default CKAN cookies for example make them http only.

**Comment by teunoknmi** [comment]
For cookies these setting in the .ini seemed to work with the logging in cookies: 
`who.secure = True
who.samesite = Strict
who.httponly = True`

Only sometimes I see a cookie like this:
![image](https://user-images.githubusercontent.com/112876766/189300891-a60ad246-a57b-4f16-8fe8-36a79f96ef19.png)

On CKAN itself it disappears quickly but a site on a subdomain it stays longer and that's where I took this screenshot from.
Any ideas where that is comming from?

**Comment by amercader** [comment, marked as answer]
The session cookie is managed by Beaker so you can use the Beaker session configuration to control it. You can use this settings, but they need to be prefixed with `beaker.session.` in your ini file (eg `beaker.session.cookie_domain`): 
https://docs.ckan.org/en/latest/maintaining/configuration.html#beaker-session-settings

These should be prefixed in the documentation, we need to fix that

**Reply by teunoknmi** [reply]
Thanks, not everything might be there, but other beaker.session.x options from https://beaker.readthedocs.io/en/latest/modules/session.html also seem work.

for example:
beaker.session.secure = true
beaker.session.httponly = true
beaker.session.samesite = Strict
